### PR TITLE
configure bond support using nmcli

### PR DESCRIPTION
**Pull request by bybai**
### The PR is for task 

https://github.com/xcat2/xcat2-task-management/issues/610

_##Feature description##_
configure bond support using nmcli

### The UT result
1.  configure nics table:
```
"byrh09","bond51!50.4.41.141",,,"bond51!bond,ens5!ethernet,ens4!ethernet",,"bond51!50_0_0_0-255_0_0_0",,,"bond51!ens5|ens4",,,
```
2. configure bond when the bond connection is not existed
```
]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: bond51 ens5@ens4
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : bond51 ens5@ens4
byrh09: [I]: create_bond_interface bondname=bond51 slave_ports=ens5,ens4 slave_type=ethernet _ipaddr=50.4.41.141
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "bond51".
byrh09: [I]: create bond connection xcat-bond51
byrh09: [I]: nmcli con add type bond con-name xcat-bond51 ifname bond51 bond.options mode=802.3ad,miimon=100 method none ipv4.method manual ipv4.addresses 50.4.41.141/8 mtu 1400
byrh09: Connection 'xcat-bond51' (f8dd01a0-3c4d-4d8f-97e1-be869c53a5d3) successfully added.
byrh09: [I]: nmcli con add type Ethernet con-name xcat-bond-slave-ens5 mtu 1400 method none ifname ens5 master xcat-bond51
byrh09: Connection 'xcat-bond-slave-ens5' (2dd338fc-3443-4062-8716-8af4a8565eb8) successfully added.
byrh09: [I]: nmcli con up xcat-bond-slave-ens5
byrh09: Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/17755)
byrh09: [I]: nmcli con add type Ethernet con-name xcat-bond-slave-ens4 mtu 1400 method none ifname ens4 master xcat-bond51
byrh09: Connection 'xcat-bond-slave-ens4' (d551897f-8e6f-4f02-971b-82742fb1c52b) successfully added.
byrh09: [I]: nmcli con up xcat-bond-slave-ens4
byrh09: Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/17756)
byrh09: [I]: nmcli con up xcat-bond51
byrh09: [I]: [bond] >> 69: bond51: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default qlen 1000
byrh09: [I]: [bond] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: [I]: [bond] >>     inet 50.4.41.141/8 brd 50.255.255.255 scope global noprefixroute bond51
byrh09: [I]: [bond] >>        valid_lft forever preferred_lft forever
byrh09: [I]: [bond] >>     inet6 fe80::5669:e83e:aa8f:24fb/64 scope link tentative noprefixroute
byrh09: [I]: [bond] >>        valid_lft forever preferred_lft forever
byrh09: postscript end....: confignetwork exited with code 0
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```
3. configure bond when the bond connection is existed
```
]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: bond51 ens5@ens4
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : bond51 ens5@ens4
byrh09: [I]: create_bond_interface bondname=bond51 slave_ports=ens5,ens4 slave_type=ethernet _ipaddr=50.4.41.141
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "bond51".
byrh09: Connection 'xcat-bond51' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/17741)
byrh09: [I]: xcat-bond51 exists, rename old xcat-bond51 to xcat-bond-xcat-bond51-tmp
byrh09: [I]: create bond connection xcat-bond51
byrh09: [I]: nmcli con add type bond con-name xcat-bond51 ifname bond51 bond.options mode=802.3ad,miimon=100 method none ipv4.method manual ipv4.addresses 50.4.41.141/8 mtu 1400
byrh09: Connection 'xcat-bond51' (72b538ff-e7db-463e-9f40-783cc31df0e2) successfully added.
byrh09: [I]: rename xcat-bond-slave-ens5 to xcat-bond-slave-ens5-tmp-1
byrh09: [I]: nmcli con add type Ethernet con-name xcat-bond-slave-ens5 mtu 1400 method none ifname ens5 master xcat-bond51
byrh09: Connection 'xcat-bond-slave-ens5' (8748282a-a70f-4650-a2ea-b03077cb1cef) successfully added.
byrh09: [I]: nmcli con up xcat-bond-slave-ens5
byrh09: Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/17747)
byrh09: [I]: rename xcat-bond-slave-ens4 to xcat-bond-slave-ens4-tmp-1
byrh09: [I]: nmcli con add type Ethernet con-name xcat-bond-slave-ens4 mtu 1400 method none ifname ens4 master xcat-bond51
byrh09: Connection 'xcat-bond-slave-ens4' (ab118652-9fff-4c82-b10f-0da466dc6199) successfully added.
byrh09: [I]: nmcli con up xcat-bond-slave-ens4
byrh09: Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/17748)
byrh09: [I]: nmcli con up xcat-bond51
byrh09: [I]: [bond] >> 69: bond51: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default qlen 1000
byrh09: [I]: [bond] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: [I]: [bond] >>     inet 50.4.41.141/8 brd 50.255.255.255 scope global noprefixroute bond51
byrh09: [I]: [bond] >>        valid_lft forever preferred_lft forever
byrh09: [I]: [bond] >>     inet6 fe80::c074:5f9d:be9d:7b4d/64 scope link tentative noprefixroute
byrh09: [I]: [bond] >>        valid_lft forever preferred_lft forever
byrh09: Connection 'xcat-bond-xcat-bond51-tmp' (e7d56555-0539-494a-b866-a1db13e210b9) successfully deleted.
byrh09: Connection 'xcat-bond-slave-ens5-tmp-1' (395668ea-682e-4a80-af77-e2e4bbe397d9) successfully deleted.
byrh09: Connection 'xcat-bond-slave-ens4-tmp-1' (90b91718-b6a4-4d8a-985a-311554506138) successfully deleted.
byrh09: postscript end....: confignetwork exited with code 0
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```
4. check results in compute node
```
]# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 42:c5:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
3: ens4: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1400 qdisc fq_codel master bond51 state UP mode DEFAULT group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
4: ens5: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1400 qdisc fq_codel master bond51 state UP mode DEFAULT group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
69: bond51: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff

]# nmcli con show
NAME                  UUID                                  TYPE      DEVICE
System ens3           a26ac740-efef-4e45-87b0-936a7236b397  ethernet  ens3
xcat-bond-slave-ens4  ab118652-9fff-4c82-b10f-0da466dc6199  ethernet  ens4
xcat-bond-slave-ens5  8748282a-a70f-4650-a2ea-b03077cb1cef  ethernet  ens5
xcat-bond51           72b538ff-e7db-463e-9f40-783cc31df0e2  bond      bond51
Wired connection 1    0ca29f07-18e0-3c16-9b2b-d0566d76d990  ethernet  --
Wired connection 2    a1686b80-8b9d-34b3-a674-c32599fda6d2  ethernet  --
```
5. regression bridge:
```
]# tabdump nics
"byrh09","br51!50.4.41.141",,,"br51!bridge,ens5!ethernet",,"br51!50_0_0_0-255_0_0_0",,,"br51!ens5",,,

]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: br51 ens5
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : br51 ens5
byrh09: [I]: create_bridge_interface_nmcli ifname=br51 _brtype=bridge _port=ens5 _pretype=ethernet _ipaddr=50.4.41.141
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "br51".
byrh09: [I]: create bridge connection xcatbr51
byrh09: [I]: nmcli con add type bridge con-name xcatbr51 ifname br51 mtu 1400
byrh09: Connection 'xcatbr51' (3fd4cb9f-29e5-4172-b7c9-5a58cb59216f) successfully added.
byrh09: [I]: create ethernet slaves connetcion xcat_br_ens5 for bridge
byrh09: [I]: nmcli con add type ethernet con-name xcat_br_ens5 ifname ens5 master br51 mtu 1400
byrh09: Connection 'xcat_br_ens5' (e4a68d91-eeaa-4551-bd27-a82e31f63cc1) successfully added.
byrh09: [I]: add ip 50.4.41.141/8 to bridge
byrh09: [I]: nmcli con up xcat_br_ens5; nmcli con up xcatbr51
byrh09: [I]: State of "br51" was "DOWN" instead of expected "UP". Wait 0 of 20 with interval 40.
byrh09: [I]: [bridge] >> 70: br51: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default qlen 1000
byrh09: [I]: [bridge] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: [I]: [bridge] >>     inet 50.4.41.141/8 brd 50.255.255.255 scope global noprefixroute br51
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: [I]: [bridge] >>     inet6 fe80::9b9c:3ad0:1e2a:45e7/64 scope link noprefixroute
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: postscript end....: confignetwork exited with code 0
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```

### CODE REVIEW

File: xCAT/postscripts/nicutils.sh [hunk 1]
```diff
@@ -1976,12 +1971,244 @@ function create_bridge_interface_nmcli {
...
+    num=1
+    xcat_slave_con_names=""
+    tmp_slave_con_names=""
+    for ifslave in $(echo "$slave_ports" | $sed -e 's/,/ /g')
+    do
+        tmp_slave_con_name=""
+        xcat_slave_con="xcat-bond-slave-"$ifslave
+        is_nmcli_connection_exist $xcat_slave_con
+        if [ $? -eq 0 ] ; then
+            tmp_slave_con_name=$xcat_slave_con"-tmp-"$num
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: seems num is not needed. And you never increase its value.

File: xCAT/postscripts/nicutils.sh [hunk 2]
```diff
@@ -1976,12 +1971,244 @@ function create_bridge_interface_nmcli {
...
+            if [ -n "$xcat_slave_con_names" ]; then
+                xcat_slave_con_names="$xcat_slave_con_names $xcat_slave_con"
+            else
+                xcat_slave_con_names=$xcat_slave_con
+            fi
+        fi
+        if [ -n "$tmp_slave_con_name" ]; then
+            is_connection_activate_intime $tmp_slave_con_name 10
+            if [ $? -eq 1 ]; then
+                $nmcli con down $tmp_slave_con_name
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: Shall it not be down con first, then check whether it is still active?

File: xCAT/postscripts/nicutils.sh [hunk 3]
```diff
@@ -1976,12 +1971,244 @@ function create_bridge_interface_nmcli {
...
+            wait_for_ifstate $bondname UP 20 10
+            if [ $? -ne 0 ]; then
+                rc=1
+            else
+                $ip address show dev $bondname| $sed -e 's/^/[bond] >> /g' | log_lines info
+            fi
+        fi
+    fi
+
+    if [ $rc -eq 1 ]; then
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: An if else with line 2143?

File: xCAT/postscripts/nicutils.sh [hunk 4]
```diff
@@ -1976,12 +1971,244 @@ function create_bridge_interface_nmcli {
...
+        # delete all bond slave and master which is created by xCAT
+        if [ -n "$xcat_slave_con_names" ]; then
+            log_info "delete connection $xcat_slave_con_names"
+            delete_bond_slaves_con "$xcat_slave_con_names"
+        fi
+        if [ -n "$tmp_slave_con_names" ]; then
+            for tmpslave in $tmp_slave_con_names
+            do  
+                slavecon=$(echo $tmpslave|awk -F"-" '{print $1"-"$2"-"$3"-"$4}')
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: It is better to store the backup con name as well as original con name?
Then, we can remove original name, then rename backup con name.

File: xCAT/postscripts/nicutils.sh (hunk 2 above)
Comment (bybai) [reply xCAT/postscripts/nicutils.sh]: Yes

File: xCAT/postscripts/nicutils.sh (hunk 4 above)
Comment (bybai) [reply xCAT/postscripts/nicutils.sh]: $(echo $tmpslave|sed 's/-tmp$//')

**Comment by bybai** [comment]
enhance configure vlan support eth->vlan->bridge UT:
1. configure nics table:
```
"byrh09","br51!50.4.41.151",,,"br51!bridge,ens5!ethernet,ens5.1!vlan",,"br51!50_0_0_0-255_0_0_0",,,"br51!ens5.1,ens5.1!ens5",,,
```
2. configure eth->vlan->bridge:
```
]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: ens5.1 ens5
byrh09: [I]: br51 ens5.1
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : ens5.1 ens5
byrh09: [I]: create_vlan_interface_nmcli ifname=ens5 vlanid=1 ipaddrs= next_nic=br51
byrh09: [I]: check parent interface ens5 whether it is managed by NetworkManager
byrh09: Connection 'xcat-vlan-ens5.1' (1b81a953-06b1-4fa2-8ce1-b15bae918c3d) successfully added.
byrh09: [I]: create NetworkManager connection for ens5.1
byrh09: [I]: [vlan] >> 168: ens5.1@ens5: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
byrh09: [I]: [vlan] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : br51 ens5.1
byrh09: [I]: create_bridge_interface_nmcli ifname=br51 _brtype=bridge _port=ens5.1 _pretype=vlan _ipaddr=50.4.41.151
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "br51".
byrh09: [I]: create bridge connection xcat-bridge-br51
byrh09: [I]: nmcli con add type bridge con-name xcat-bridge-br51 ifname br51 mtu 1400
byrh09: Connection 'xcat-bridge-br51' (f7459ae9-5dcd-4acc-b5b4-44993c6700c5) successfully added.
byrh09: [I]: create vlan slaves connetcion xcat-vlan-ens5.1 for bridge
byrh09: [I]: nmcli con mod xcat-vlan-ens5.1 master br51 mtu 1400
byrh09: [I]: add ip 50.4.41.151/8 to bridge
byrh09: [I]: nmcli con up xcat-bridge-br51
byrh09: Connection successfully activated (master waiting for slaves) (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/21539)
byrh09: [I]: nmcli con up xcat-vlan-ens5.1
byrh09: Connection successfully activated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/21540)
byrh09: [I]: State of "br51" was "DOWN" instead of expected "UP". Wait 0 of 20 with interval 40.
byrh09: [I]: [bridge] >> 169: br51: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default qlen 1000
byrh09: [I]: [bridge] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: [I]: [bridge] >>     inet 50.4.41.151/8 brd 50.255.255.255 scope global noprefixroute br51
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: [I]: [bridge] >>     inet6 fe80::be6c:503f:3e94:7ebb/64 scope link noprefixroute
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: postscript end....: confignetwork exited with code 0
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```
3. check results in CN
```
]# nmcli con show
NAME              UUID                                  TYPE      DEVICE
System ens3       a26ac740-efef-4e45-87b0-936a7236b397  ethernet  ens3
ens5              0e3f33de-a0f2-4911-bc74-ca9baf9736d7  ethernet  ens5
xcat-bridge-br51  f7459ae9-5dcd-4acc-b5b4-44993c6700c5  bridge    br51
xcat-vlan-ens5.1  1b81a953-06b1-4fa2-8ce1-b15bae918c3d  vlan      ens5.1
ens4              6dab9d60-7d01-4782-838a-5236bb5f51a5  ethernet  --

]# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 42:c5:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
3: ens4: <BROADCAST,MULTICAST> mtu 1500 qdisc fq_codel state DOWN mode DEFAULT group default qlen 1000
    link/ether 42:20:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
4: ens5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
168: ens5.1@ens5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue master br51 state UP mode DEFAULT group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
169: br51: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
```

File: xCAT/postscripts/confignetwork [hunk 5]
```diff
@@ -623,13 +628,13 @@ errorcode=0
 utolcmd="sed -e y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
 
 #back up all network interface configure files
-nwdirbak=$nwdir".xcatbak"
+nwdirbak="/tmp/"
```
Comment (zet809) [inline xCAT/postscripts/confignetwork]: Why do this changing?

File: xCAT/postscripts/nicutils.sh [hunk 6]
```diff
@@ -1976,12 +1993,263 @@ function create_bridge_interface_nmcli {
...
+    tmp_slave_con_names=""
+    for ifslave in $xcat_slave_ports
+    do
+        tmp_slave_con_name=""
+        xcat_slave_con="xcat-bond-slave-"$ifslave
+        is_nmcli_connection_exist $xcat_slave_con
+        if [ $? -eq 0 ] ; then
+            is_connection_activate_intime $xcat_slave_con 1
+            if [ $? -eq 1 ]; then
+                $nmcli con down $xcat_con_name
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: Why do xcat_con_name? Shall be xcat_slave_con?

File: xCAT/postscripts/confignetwork (hunk 5 above)
Comment (bybai) [reply xCAT/postscripts/confignetwork]: "/tmp" is not a good place to save xcatbackup directory, so will keep it as original directory.

**Comment by zet809** [comment]
Great to me, thx @bybai

**Pull request merged**